test/unit/lib/cms/aws_credentials_test.rb

```diff
@@ -0,0 +1,108 @@
+  end
+
+  teardown do
+    CMS::AwsCredentials.instance.instance_variable_set('@credentials', nil)
+  end
+
+  test 'returns IAM credentials when available' do
+    use_iam
+    CMS::AwsCredentials.instance.expects(:sts_credentials).never
+
+    assert_equal iam_auth_params, CMS::AwsCredentials.instance.credentials
+  end
+
+  test 'returns STS credentials when available' do
+    use_sts
+    CMS::AwsCredentials.instance.expects(:sts_credentials).returns(sts_credentials)
+
+    assert_equal({ credentials: sts_credentials }, CMS::AwsCredentials.instance.credentials)
+  end
+
+  test 'returns IAM credentials when both authentication types are available' do
+    use_full_params
+    CMS::AwsCredentials.instance.expects(:sts_credentials).never
+
+    assert_equal iam_auth_params, CMS::AwsCredentials.instance.credentials
+  end
+
+  test 'raises an error if not enough credential params are present' do
+    CMS::S3.stubs(:config).returns(full_params.except(:role_arn, :access_key_id))
+
+    assert_raises(CMS::AwsCredentials::AuthenticationTypeError) do
+      CMS::AwsCredentials.instance.credentials
+    end
+  end
+
+  test 'is a singleton' do
+    assert CMS::AwsCredentials <= Singleton
+  end
+
+  test 'raises an error if the web_identity_token_file does not exist' do
+    use_sts
+
+    assert_raises(Aws::Errors::MissingWebIdentityTokenFile) do
+      CMS::AwsCredentials.instance.credentials
+    end
+  end
+
+  test 'STS credentials are created with correct parameters' do
+    use_sts
+    File.stubs(:exist?).with(full_params[:web_identity_token_file]).returns(true)
+
+    Aws::AssumeRoleWebIdentityCredentials.expects(:new).with(sts_auth_params).returns(sts_credentials)
+    assert_equal sts_credentials, CMS::AwsCredentials.instance.send(:sts_credentials)
+  end
+
+  private
+
+  def full_params
+    {
+      access_key_id: 'my_access_key_id',
+      secret_access_key: 'my_secret_access_key',
+      web_identity_token_file: '/path/to/token',
+      role_arn: 'my_role_arn',
+      role_session_name: 'my_role_session_name',
+      region: 'my-region'
+    }
+  end
+
+  def iam_auth_params
+    full_params.slice(*CMS::AwsCredentials::IAM_KEYS)
+  end
+
+  def sts_auth_params
+    full_params.slice(:region, *CMS::AwsCredentials::STS_KEYS)
+  end
+
+  def use_iam
+    CMS::S3.stubs(:config).returns(full_params.except(*CMS::AwsCredentials::STS_KEYS))
+  end
+
+  def use_sts
+    CMS::S3.stubs(:config).returns(full_params.except(*CMS::AwsCredentials::IAM_KEYS))
+  end
+
+  def use_full_params
+    CMS::S3.stubs(:config).returns(full_params)
+  end
+
+  def sts_credentials
+    @sts_credentials ||= Aws::Credentials.new(nil, nil)
+  end
+
+  def with_fake_token
+    FakeFS do
+      FakeFS::FileSystem.clone(Gem.loaded_specs['aws-partitions'].full_gem_path)
+      FileUtils.mkdir_p(File.dirname(full_params[:web_identity_token_file]))
+      File.write(full_params[:web_identity_token_file], "akhsdkjfskdfjkdshfkjd")
```

Comment:
`full_params[:web_identity_token_file]` is called twice in here. It would be caught by reek in CodeClimate but we [disabled it](https://github.com/3scale/porta/blob/master/.codeclimate.yml#L54) for test files.

Reply:
I guess so, but we are not after efficiency here and I believe it is well clear just access to the hash. Extracting a variable will not improve readability.

Reply:
I agree it's clear enough as it is but I don't think extracting it would impact on the readability, as this is a small method (as they should be). And just to be clear, this is not a blocker at all for this PR. We are skipping CodeClimate checks from the test files already, so we're not bound to reek/rubocop/something suggestions.

My only concern is with setting standards for the codebase. Was this double method call in a non test file, we'd get complains from CodeClimate, which would lead us to resolve them or use magic comments.
I think the main point here is to come up with an agreement (maybe guidelines 🤔) for this repo's contributors, so that we can avoid the discussions that go down to personal preference.

Again, this PR is more than ready to go 🚀

Reply:
In testing I would prioritize readability and simplicity. In production reliability, maintainability and efficiency. Although first two should still apply.

Reply:
btw this method is not needed anymore, it was only used in the deleted test case